api.c

```diff
@@ -0,0 +1,1055 @@
+    }
+    str->value[len_str] = c;
+    str->value[len_str+1] = '\0';
+}
+
+static void push_back(arr_t *arr, str_t str) {
+    arr->items = (str_t *) realloc(arr->items, sizeof(str_t) * (arr->size + 2));
+
+    str_t toadd;
+    init_str(&toadd, 0);
+    toadd.value = strdup(str.value);
+    toadd.size = str.size;
+
+    arr->items[arr->size] = toadd;
+    arr->size++;
+}
+
+static void init_str_arr(arr_t *arr) {
+    str_t null;
+    init_str(&null, 0);
+    arr->items = (str_t *) malloc(0);
+    arr->items[0] = null;
+    arr->size = 0;
+}
+
+/**
+ * Arguments parser based on "stringtoargcargv.cpp" by Bernhard Eder
+ * @see https://web.archive.org/web/20121030075237/http://bbgen.net/blog/2011/06/string-to-argc-argv
+ */
+static int parse_args(arg_data *data, char str[]) {
+    char c;
+    int i = 0;
+    str_t curr_arg;
+    arr_t out_args;
+    init_str(&curr_arg, 20);
+    init_str_arr(&out_args);
+    enum State {
+        InArg,
+        InArgQuote,
+        OutOfArg,
+    };
+    enum State curr_state = OutOfArg;
+    char curr_quote = '\0';
+
+    while ((c = str[i]) != '\0') {
+        if (c == '\"' || c == '\'') { // Quote
+            switch (curr_state) {
+                case OutOfArg:
+                    free(curr_arg.value);
+                    init_str(&curr_arg, 20);
+                case InArg:
+                    curr_state = InArgQuote;
+                    curr_quote = c;
+                    break;
+                case InArgQuote:
+                    if (c == curr_quote)
+                        curr_state = InArg;
+                    else
+                        str_push_c(&curr_arg, c);
+                    break;
+            }
+        } else
+        if (c == ' ' || c == '\t') { // Whitespace
+            switch(curr_state) {
+                case InArg:
+                    push_back(&out_args, curr_arg);
+                    curr_state = OutOfArg;
+                    break;
+                case InArgQuote:
+                    str_push_c(&curr_arg, c);
+                    break;
+                case OutOfArg:
+                    break;
+            }
+        } else {
+            switch(curr_state) {
+                case InArg:
+                case InArgQuote:
+                    str_push_c(&curr_arg, c);
+                    break;
+                case OutOfArg:
+                    free(curr_arg.value);
+                    init_str(&curr_arg, 20);
+                    str_push_c(&curr_arg, c);
+                    curr_state = InArg;
+                    break;
+            }
+        };
+        i++;
+    }
+
+    if (curr_state == InArg) {
+        push_back(&out_args, curr_arg);
+    } else if (curr_state == InArgQuote) {
+        return 1;
+    }
+
+    size_t argc = out_args.size;
+    char **argv = str_destruct(out_args);
+
+    free(curr_arg.value);
+    free(out_args.items);
+
+    data->argc = argc;
+    data->argv = argv;
+
+    return 0;
+}
+
+void *listen_api(void* _) {
```

Comment:
what's with the void pointers here?

Reply:
A `void*` function is required by `pthread_create`, although I could remove the `void* _` part.

Reply:
Oh this is called by pthread, nvm then. But I believe (void*) works too.